lib/reversetunnel/cache.go

```diff
@@ -60,25 +59,23 @@ func NewHostCertificateCache(keygen sshca.Authority, authClient auth.ClientI) (*
 // Multiple callers can arrive and generate a host certificate at the same time.
 // This is a tradeoff to prevent long delays here due to the expensive
 // certificate generation call.
-func (c *certificateCache) GetHostCertificate(addr string) (ssh.Signer, error) {
+func (c *certificateCache) GetHostCertificate(addr string, additionalPrincipals []string) (ssh.Signer, error) {
```

Comment:
what if the list of additional principals for the certificate has changed? you might want to invalidate the cache then. Kube forwarder solves this problem by making principals to be part of the key:

https://github.com/gravitational/teleport/blob/master/lib/kube/proxy/forwarder.go#L194

Could something similar be relevant in your use-case as well?